test/test-tools.go

```diff
@@ -157,6 +157,51 @@ func AssertNotContains(t *testing.T, haystack string, needle string) {
 	}
 }
 
+// AssertMetricEquals determines whether the value held by a prometheus Collector
+// (e.g. Gauge, Counter, CounterVec, etc) is equal to the expected integer.
+// In the case of vector collectors, it collects and sums the values from all
+// metrics in the vector prior to comparison; this is so that users can easily
+// make assertions about one of many field dimensions (e.g. for a CounterVec with
+// fields "host" and "valid", being able to assert that two "valid": "true"
+// incremets occurred, without caring which host did each increment). To make
+// assertions about subsets of a vector's metrics, use the MetricVec's .CurryWith
+// or .GetMetricWith methods. Only works for integer metrics (Counters and Gauges),
+// does not work for Histograms.
+func AssertMetricWithLabelsEquals(t *testing.T, c prometheus.Collector, l prometheus.Labels, expected int) {
+	ch := make(chan prometheus.Metric)
+	done := make(chan struct{})
+	go func() {
+		c.Collect(ch)
+		close(done)
```

Comment:
Why do we have a separate `done` channel to close rather than closing `ch`? We know `c.Collect` will not write anything else to it.

Reply:
Because it seemed more obviously correct to have a wholly separate case for being done, rather than having a nested conditional inside the `m <- ch` case which checks for receiving `nil`. I didn't want to make the assumption that `.Collect()` would _never_ send a `nil` metric through the channel.

Reply:
That's reasonable! Though note that you wouldn't do this with a `nil` check, but would instead use the `x, ok := <- ch` form, and check `ok`.